superset-frontend/src/explore/components/controls/AnnotationLayerControl/AnnotationLayer.jsx

```diff
@@ -298,73 +320,149 @@ class AnnotationLayer extends React.PureComponent {
+      });
+
+      const { result, count } = json;
+
+      const layersObj = {};
+      result.forEach(layer => {
+        layersObj[layer.id] = {
+          value: layer.id,
+          label: layer.name,
+        };
+      });
+
+      const layersArray = Object.values(layersObj);
+
+      this.setState(prevState => ({
+        valueOptions: { ...prevState.valueOptions, ...layersObj },
+      }));
+
+      return {
+        data: layersArray,
+        totalCount: count,
+      };
+    }
+
+    if (requiresQuery(sourceType)) {
+      const queryParams = rison.encode({
+        filters: [
+          {
+            col: 'id',
+            opr: 'chart_owned_created_favored_by_me',
+            value: true,
+          },
+        ],
+        order_column: 'slice_name',
+        order_direction: 'asc',
+        page,
+        page_size: pageSize,
+      });
+      const { json } = await SupersetClient.get({
+        endpoint: `/api/v1/chart/?q=${queryParams}`,
+      });
+
+      const { result, count } = json;
+      const registry = getChartMetadataRegistry();
+
+      const chartsObj = result
+        .filter(chart => {
+          const metadata = registry.get(chart.viz_type);
+          return metadata && metadata.canBeAnnotationType(annotationType);
+        })
+        .reduce((acc, chart) => {
+          acc[chart.id] = {
+            value: chart.id,
+            label: chart.slice_name,
+            slice: {
+              ...chart,
+              data: {
+                ...chart.form_data,
+                groupby: chart.form_data.groupby?.map(column =>
+                  getColumnLabel(column),
+                ),
+              },
             },
-          ],
-          order_column: 'slice_name',
-          order_direction: 'asc',
-          page: 0,
-          page_size: FILTER_OPTIONS_LIMIT,
-        });
-        SupersetClient.get({
-          endpoint: `/api/v1/chart/?q=${queryParams}`,
-        }).then(({ json }) => {
-          const registry = getChartMetadataRegistry();
-          this.setState({
-            isLoadingOptions: false,
-            valueOptions: json.result
-              .filter(x => {
-                const metadata = registry.get(x.viz_type);
-                return metadata && metadata.canBeAnnotationType(annotationType);
-              })
-              .map(x => ({
-                value: x.id,
-                label: x.slice_name,
+          };
+          return acc;
+        }, {});
+
+      const chartsArray = Object.values(chartsObj);
+
+      this.setState(prevState => ({
+        valueOptions: { ...prevState.valueOptions, ...chartsObj },
+      }));
+
+      return {
+        data: chartsArray,
+        totalCount: count,
+      };
+    }
+
+    return {
+      data: [],
+      totalCount: 0,
+    };
+  };
+
+  fetchAppliedChart() {
+    const { value, valueOptions, sourceType } = this.state;
+    const id = value;
+    if (id && !valueOptions[id] && requiresQuery(sourceType)) {
```

Comment:
You could move this conditional to the `didUpdate` so you don't call this function is not met?